lib/course-db.js

```diff
@@ -158,49 +158,72 @@ function checkInfoValid(idName, info, infoFile, courseInfo, logger, cache) {
-            jsonLoad.readInfoJSON(infoFile, schema, function(err, info) {
-                if (err && err.code && err.path && (err.code === 'ENOTDIR') && err.path === infoFile) {
-                    // In a previous version of this code, we'd pre-filter
-                    // all files in the parent directory to remove anything
-                    // that may have accidentally slipped in, like .DS_Store.
-                    // However, that resulted in a huge number of system calls
-                    // that got really slow for large directories. Now, we'll
-                    // just blindly try to read a file from the directory and assume
-                    // that if we see ENOTDIR, that means the directory was not
-                    // in fact a directory.
-                    callback(null);
-                    return;
-                }
-                if (ERR(err, callback)) return;
-                jsonLoad.validateOptions(info, infoFile, optionSchemaPrefix, schemas, function(err, info) {
-                    if (ERR(err, callback)) return;
-                    info[idName] = dir;
-                    info.directory = dir;
-                    err = checkInfoValid(idName, info, infoFile, courseInfo, logger, cache);
-                    if (ERR(err, callback)) return;
-                    if (info.disabled) {
-                        callback(null);
+    let walk = function(relativeDir, callback) {
+        fs.readdir(path.join(parentDir, relativeDir), function(err, files) {
+            if (ERR(err, callback)) return;
+
+            async.map(files, function(dir, callback) {
+                var infoFile = path.join(parentDir, relativeDir, dir, infoFilename);
+                jsonLoad.readInfoJSON(infoFile, schema, function(err, info) {
+                    if (err && err.code && err.path && (err.code === 'ENOTDIR') && err.path === infoFile) {
+                        // In a previous version of this code, we'd pre-filter
+                        // all files in the parent directory to remove anything
+                        // that may have accidentally slipped in, like .DS_Store.
+                        // However, that resulted in a huge number of system calls
+                        // that got really slow for large directories. Now, we'll
+                        // just blindly try to read a file from the directory and assume
+                        // that if we see ENOTDIR, that means the directory was not
+                        // in fact a directory.
+                        callback(null, 0);
                         return;
                     }
-                    info = _.defaults(info, defaultInfo);
-                    db[dir] = info;
-                    return callback(null);
+                    if (recursive && err && err.code && (err.code == 'ENOENT') && err.path === infoFile) {
+                        /* Try to recurse into the directory if we don't find an info json file. */
```

Comment:
I like the way you are clever about only recursing if we don't have an info file at this level.